crates/core/src/metadata_store/providers/objstore/object_store_version_repository.rs

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) 2024 - Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::sync::Arc;
+
+use bytes::Bytes;
+use futures::TryStreamExt;
+use object_store::aws::AmazonS3Builder;
+use object_store::aws::S3ConditionalPut::ETagMatch;
+use object_store::path::Path;
+use object_store::{Error, ObjectStore, PutMode, PutOptions, PutPayload};
+
+use crate::metadata_store::providers::objstore::version_repository::{
+    VersionRepository, VersionRepositoryError,
+};
+use restate_types::config::MetadataStoreClient;
+use restate_types::Version;
+
+#[derive(Debug)]
+pub(crate) struct ObjectStoreVersionRepository {
+    object_store: Arc<dyn ObjectStore>,
+}
+
+impl ObjectStoreVersionRepository {
+    pub(crate) fn from_configuration(configuration: MetadataStoreClient) -> Self {
+        let store = match configuration {
+            // TODO: really configure this
+            MetadataStoreClient::ObjectStore { address, bucket } => {
+                let mut builder = AmazonS3Builder::from_env()
+                    .with_bucket_name(bucket)
+                    .with_endpoint(address)
+                    .with_conditional_put(ETagMatch)
+                    .with_allow_http(true);
+
+                builder = builder
+                    .with_access_key_id("minioadmin")
+                    .with_secret_access_key("minioadmin");
+
+                builder.build().expect("Failed to create AWS Amazon S3")
+            }
+            _ => unreachable!(),
+        };
+        Self {
+            object_store: Arc::new(store),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl VersionRepository for ObjectStoreVersionRepository {
+    async fn get_latest_version(&self) -> Result<Option<Version>, VersionRepositoryError> {
+        let mut stream = self.object_store.list(None);
+
+        // TODO: we should switch to a create internal Version
+        let mut current = 0;
+        while let Some(version) = stream
+            .try_next()
+            .await
+            .map_err(|e| VersionRepositoryError::Network(e.into()))?
+        {
+            if let Some(name) = version.location.filename() {
+                if !name.starts_with("v") {
+                    continue;
+                }
+                if let Ok(v) = name[1..].parse::<u32>() {
+                    if v > current {
+                        current = v;
+                    }
+                }
+            }
+        }
```

Comment:
I guess this operation should motivate us to clean up older version so that the iteration won't grow as time passes.

Reply:
I think we can do this without LIST operations but rather using object versioning.